### PR TITLE
Username generator

### PR DESCRIPTION
**Pull request by kspearrin**
## Type of change

- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective

Add support for username generation.

Asana task: https://app.asana.com/0/1200578627114106/1201532987437763

## Code changes

- Restructure password generator component to be a more generic "generator" with options for passwords and usernames.
- Add username generator service for generating usernames of various types.
- Expand state service to store generator and username generation settings.
- Add support for generating usernames to item add/edit component.

Note: I plan to follow up at some point and rename the password generator component to just "generator".

## Testing requirements

1. Test existing password generation options from generator.
2. Test new username generation options from generator.
3. Test integration of username and password generation into item add/edit page.

## Before you submit

- [x] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [x] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

**Comment by differsthecat** [comment]
Looks great, my only concern was making sure a user's generator settings for a given type were save and retrieved when switching types, but it looks like that is handled!

**Review by differsthecat — APPROVED** [review]

File: angular/src/components/add-edit.component.ts [hunk 1]
```diff
@@ -425,6 +426,27 @@ export class AddEditComponent implements OnInit {
     return true;
   }
 
+  async generateUsername(): Promise<boolean> {
+    if (
+      this.cipher.login != null &&
+      this.cipher.login.username != null &&
+      this.cipher.login.username.length
```
Comment (Hinton) [inline angular/src/components/add-edit.component.ts:433]:
```suggestion
this.cipher.login?.username?.length
```

File: common/src/abstractions/state.service.ts [hunk 2]
```diff
@@ -265,6 +265,10 @@ export abstract class StateService<T extends Account = Account> {
   ) => Promise<void>;
   getPasswordGenerationOptions: (options?: StorageOptions) => Promise<any>;
   setPasswordGenerationOptions: (value: any, options?: StorageOptions) => Promise<void>;
+  getUsernameGenerationOptions: (options?: StorageOptions) => Promise<any>;
```
Comment (Hinton) [inline common/src/abstractions/state.service.ts:268]: Could we add a proper return type for this?

File: common/src/abstractions/usernameGeneration.service.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+export abstract class UsernameGenerationService {
+  generateUsername: (options: any) => Promise<string>;
+  generateWord: (options: any) => Promise<string>;
+  generateSubaddress: (options: any) => Promise<string>;
+  generateCatchall: (options: any) => Promise<string>;
+  getOptions: () => Promise<any>;
+  saveOptions: (options: any) => Promise<void>;
```
Comment (Hinton) [inline common/src/abstractions/usernameGeneration.service.ts:7]: Same here, `any` are a pain to know what they expect.

File: common/src/abstractions/state.service.ts (hunk 2 above)
Comment (kspearrin) [reply common/src/abstractions/state.service.ts:268]: We could, but I feel like we need to do more refactoring work for all generation options that currently exist as well to keep things consistent.  Maybe a tech debt task?

Comment (Hinton) [reply common/src/abstractions/state.service.ts:268]: I think we could quite easily add an interface for the option, and just use `Partial<UsernameGeneratorOptions>` for all the functions that expects it. And leave password as a tech debt.

But I leave that up to you to decide.

**Comment by Hinton** [comment]
@differsthecat I just took a very quick glance at this so leaving the approval to you :).

**Comment by kspearrin** [comment]
Merging since @differsthecat had already approved previously. I'll add a new tech debt item item regarding generation options typing.

**Pull request merged**